File: apps/ios/App/ViewModels/LoansViewModel.swift
import Foundation
import Observation
import Factory
import Domain
import Data
import Supabase

/// Base/display currency -- loans are always created in the base currency
/// (matches web: `AddLoan` inserts with `currency: base`, no per-loan
/// currency picker exists), matching Android's LoansViewModel.kt /
/// InvestmentsViewModel.swift's own established simplification.
private let BASE_CURRENCY = "INR"
private let NON_INVESTMENT_TYPES: Set<String> = ["stocks", "mutual_funds", "demat"]

/// Ported from apps/web/app/loans/page.tsx per docs/mobile/screen-specs/
/// loans.md (task #27/#42). Was a broken placeholder before this pass
/// (2026-08-06): non-optional access on now-nullable Loan fields (a real
/// crash risk once LoansRepository.swift was fixed to match web's actually-
/// nullable columns), a `"Loans & Recurring"` title that appears to be
/// invented drift merging with a separate, unbuilt "Recurring" nav item
/// (same class of bug as the earlier-fixed Goals/Cashflow merge -- see
/// LoansView.swift), and a placeholder `"Day \(dueDay)"` next-due string.
/// Rewritten to mirror Android's LoansViewModel.kt field-for-field: grouped
/// list totals, real create(), and driven entirely by
/// LoansRepository.watchLoans() (real db.watch(), not a one-shot list) so
/// this doesn't hit the list-staleness bug fixed for Goals/Budgets this
/// same pass (2026-08-06) -- see AUDIT_HISTORY.md.
@Observable
@MainActor
public final class LoansViewModel {
    @ObservationIgnored
    @Injected(\.loansRepository) private var loansRepository
    @ObservationIgnored
    @Injected(\.ledgerRepository) private var ledgerRepository
    @ObservationIgnored
    @Injected(\.authRepository) private var authRepository

    public struct FundingAccountOption: Identifiable, Equatable, Sendable {
        public let id: String
        public let name: String
        public let isCreditCard: Bool
    }

    public struct LoanUiModel: Identifiable, Equatable, Sendable {
        public let id: String
        public let lender: String
        public let active: Bool
        public let rangeOrRate: String
        public let paidCountText: String
        public let progress: Double // 0...1, or 0 if no tenure
        public let hasTenure: Bool
        public let principalFormatted: String
        public let emiFormatted: String
    }

    public var loans: [LoanUiModel] = []
    public var totalEmiFormatted: String = formatMoney(0, BASE_CURRENCY)
    public var fundingAccounts: [FundingAccountOption] = []

    private var observeTask: Task<Void, Never>?

    public init() {
        start()
    }

    /// Idempotent -- guarded by `observeTask` alone, safe to call from every
    /// `.onAppear`.
    public func start() {
        guard observeTask == nil else { return }
        observeTask = Task { [weak self] in
            guard let self else { return }
            guard let userId = await self.resolveUserId() else { return }
            do {
                let stream = try await self.loansRepository.watchLoans(userId: userId)
                for try await dbLoans in stream {
                    await self.rebuild(dbLoans)
                }
            } catch {
                print("Failed to observe loans: \(error)")
            }
        }
    }

    public func cancel() {
        observeTask?.cancel()
        observeTask = nil
    }

    private func rebuild(_ dbLoans: [Loan]) async {
        let balances = (try? await ledgerRepository.accountBalances()) ?? []
        let rates = (try? await ledgerRepository.rates()) ?? { _, _ in 1.0 }

        fundingAccounts = balances
            .filter { !NON_INVESTMENT_TYPES.contains($0.account.type) }
            .sorted { ($0.account.type == "credit_card" ? 0 : 1) < ($1.account.type == "credit_card" ? 0 : 1) }
            .map { FundingAccountOption(id: $0.account.id, name: $0.account.name, isCreditCard: $0.account.type == "credit_card") }

        let today = isoToday()
        var totalEmiBase: Int64 = 0
        loans = dbLoans.map { l in
            let tenure = l.tenureMonths ?? 0
            let paid = paidCount(l, today)
            let remaining = tenure > 0 ? max(0, tenure - paid) : nil
            let closed = tenure > 0 && remaining == 0
            let range = loanRange(l.startDate, tenure)
            let cur = l.currency.isEmpty ? BASE_CURRENCY : l.currency
            if let emi = l.emiAmount {
                totalEmiBase += cur == BASE_CURRENCY ? emi : ((try? convert(money(emi, cur), to: BASE_CURRENCY, rate: rates(cur, BASE_CURRENCY)).amount) ?? emi)
            }
            return LoanUiModel(
                id: l.id,
                lender: (l.lender?.isEmpty == false) ? l.lender! : "Loan",
                active: !closed,
                rangeOrRate: range ?? (l.interestRate.map { "\(formatRate($0))% p.a." } ?? "—"),
                paidCountText: tenure > 0 ? "\(paid) / \(tenure) paid" : "\(paid) paid",
                progress: tenure > 0 ? min(1.0, max(0.0, Double(paid) / Double(tenure))) : 0,
                hasTenure: tenure > 0,
                principalFormatted: formatMoney(l.principal, cur),
                emiFormatted: l.emiAmount.map { formatMoney($0, cur) } ?? (l.rateType == "variable" ? "Varies" : "—")
            )
        }
        totalEmiFormatted = formatMoney(totalEmiBase, BASE_CURRENCY)
    }

    /// Matches web's `AddLoan.save()`: requires a lender name or a
    /// principal; EMI is caller-supplied (auto-calculated by the screen
    /// for fixed-rate loans via `emiFromPrincipal`, nil for variable).
    public func create(
        lender: String, principalMajorText: String, emiMajorText: String?, interestRateText: String,
        tenureText: String, startDate: String?, dueDayText: String, autoMarkPaid: Bool, rateType: String,
        fundingAccountId: String?, alertTimeUtc: String
    ) async -> String? {
        let trimmedLender = lender.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmedLender.isEmpty && principalMajorText.isEmpty { return "Enter a lender or a loan amount." }
        guard let userId = await resolveUserId() else { return "Couldn't determine the current user." }
        do {
            let principalMinor = fromMajor(Double(principalMajorText) ?? 0, BASE_CURRENCY).amount
            let dueDay = Int(dueDayText).map { min(31, max(1, $0)) }
            try await loansRepository.create(
                userId: userId,
                input: NewLoanInput(
                    lender: trimmedLender, currency: BASE_CURRENCY, principal: principalMinor,
                    emiAmount: emiMajorText.flatMap { Double($0) }.map { fromMajor($0, BASE_CURRENCY).amount },
                    interestRate: Double(interestRateText) ?? 0, tenureMonths: Int(tenureText),
                    startDate: startDate, emiDueDay: dueDay, autoMarkPaid: autoMarkPaid, rateType: rateType,
                    fundingAccountId: fundingAccountId, alertTimeUtc: alertTimeUtc
                )
            )
            return nil
        } catch {
            return "Couldn't add the loan: \(error.localizedDescription)"
        }
    }

    private func resolveUserId() async -> String? {
        if let existing = authRepository.currentUserId { return existing }
        return try? await authRepository.ensureUser()
    }
}

/// Effective paid-EMI count for a loan row (manual marks union auto-marked
/// past-due) -- matches web's page-local `paidCount()` exactly, and
/// Android's LoansViewModel.kt `paidCount()`.
func paidCount(_ l: Loan, _ todayIso: String) -> Int {
    let tenure = l.tenureMonths ?? 0
    var manual = parseManualPaid(l.emiPayments)
    if manual.isEmpty, let n = l.emisPaid, n > 0 { manual = Array(1...n) }
    return effectivePaidEmis(manual: manual, totalEmis: tenure, autoMark: l.autoMarkPaid, startIso: l.startDate, dueDay: l.emiDueDay, asOfIso: todayIso).count
}

/// Parses the `emi_payments` JSON map `{ "1": "2026-01-05", "2": "" }` into
/// the list of EMI numbers with a non-blank paid-on date.
func parseManualPaid(_ json: String?) -> [Int] {
    guard let json, !json.isEmpty, let data = json.data(using: .utf8) else { return [] }
    guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return [] }
    return obj.compactMap { key, value -> Int? in
        let str = (value as? String) ?? ""
        guard !str.isEmpty, let n = Int(key) else { return nil }
        return n
    }
}

/// "Mar '26 - Nov '26" from a start date + tenure in months -- matches
/// web's `loanRange()` / Android's `loanRange()`.
func loanRange(_ startIso: String?, _ tenure: Int) -> String? {
    guard let startIso, !startIso.isEmpty, tenure > 0 else { return nil }
    let s = String(startIso.prefix(10))
    let parts = s.split(separator: "-")
    guard parts.count == 3, let y = Int(parts[0]), let m = Int(parts[1]), let d = Int(parts[2]) else { return nil }
    var cal = Calendar(identifier: .gregorian)
    cal.timeZone = TimeZone(identifier: "UTC")!
    var comps = DateComponents(); comps.year = y; comps.month = m; comps.day = d
    guard let start = cal.date(from: comps), let end = cal.date(byAdding: .month, value: tenure, to: start) else { return nil }
    let fmt = DateFormatter()
    fmt.dateFormat = "MMM ''yy"
    fmt.timeZone = TimeZone(identifier: "UTC")
    fmt.locale = Locale(identifier: "en_US")
    return "\(fmt.string(from: start)) \u{2013} \(fmt.string(from: end))"
}

func formatRate(_ r: Double) -> String {
    r == r.rounded() ? String(Int64(r)) : String(r)
}

// formatMoney moved to App/Components/MoneyFormat.swift — one formatter,
// hide-amounts aware, fraction digits from minorUnits(currency).

File: apps/android/app/src/main/java/com/sanvya/app/ui/loans/LoansViewModel.kt
package com.sanvya.app.ui.loans

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sanvya.app.data.auth.AuthRepository
import com.sanvya.app.data.repository.Loan
import com.sanvya.app.data.repository.LedgerRepository
import com.sanvya.app.data.repository.LoansRepository
import com.sanvya.app.data.repository.NewLoanInput
import com.sanvya.app.domain.finance.effectivePaidEmis
import com.sanvya.app.domain.money.convert
import com.sanvya.app.domain.money.fromMajor
import com.sanvya.app.domain.money.money
import com.sanvya.app.domain.money.toMajor
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import java.text.NumberFormat
import java.time.LocalDate
import java.util.Currency
import java.util.Locale

/** Base/display currency -- loans are always created in the base currency
 * (matches web: `AddLoan` inserts with `currency: base`, no per-loan
 * currency picker exists). Hardcoded "INR" matching Dashboard/
 * Investments' own established simplification. */
private const val BASE_CURRENCY = "INR"

private val NON_INVESTMENT_TYPES = setOf("stocks", "mutual_funds", "demat")

data class FundingAccountOption(val id: String, val name: String, val isCreditCard: Boolean)

data class LoanUiModel(
    val id: String,
    val lender: String,
    val active: Boolean,
    val rangeOrRate: String,
    val paidCountText: String,
    val progress: Double, // 0..1, or 0 if no tenure
    val hasTenure: Boolean,
    val principalFormatted: String,
    val emiFormatted: String,
)

/**
 * Ported from apps/web/app/loans/page.tsx per docs/mobile/screen-specs/
 * loans.md (task #27). Was read-only (watchLoans() only, constructor-
 * injected dead code, no consuming Screen, no nav route -- same
 * "reported DONE, actually never real" pattern as Budgets/Goals/
 * Investments' old ViewModels) before this pass (2026-08-06); replaced
 * with a real KoinComponent/by-inject() one, removed the now-superseded
 * LoanUiModel from ui/UiModels.kt.
 */
class LoansViewModel : ViewModel(), KoinComponent {
    private val loansRepository: LoansRepository by inject()
    private val ledgerRepository: LedgerRepository by inject()
    private val authRepository: AuthRepository by inject()

    private val _loans = MutableStateFlow<List<LoanUiModel>>(emptyList())
    val loans: StateFlow<List<LoanUiModel>> = _loans

    private val _totalEmiFormatted = MutableStateFlow(formatMoney(0L))
    val totalEmiFormatted: StateFlow<String> = _totalEmiFormatted

    private val _fundingAccounts = MutableStateFlow<List<FundingAccountOption>>(emptyList())
    val fundingAccounts: StateFlow<List<FundingAccountOption>> = _fundingAccounts

    init {
        viewModelScope.launch {
            val userId = authRepository.currentUserId.value ?: return@launch
            combine(
                loansRepository.watchLoans(userId),
                ledgerRepository.watchAccounts(),
                ledgerRepository.watchRates(),
            ) { loans, accounts, rates -> Triple(loans, accounts, rates) }.collect { (loans, accounts, rates) ->
                _fundingAccounts.value = accounts
                    .filter { it.type !in NON_INVESTMENT_TYPES }
                    .sortedByDescending { it.type == "credit_card" }
                    .map { FundingAccountOption(it.id, it.name, it.type == "credit_card") }

                val today = LocalDate.now().toString()
                var totalEmiBase = 0L
                val uis = loans.map { l ->
                    val tenure = l.tenureMonths ?: 0
                    val paid = paidCount(l, today)
                    val remaining = if (tenure > 0) maxOf(0, tenure - paid) else null
                    val closed = tenure > 0 && remaining == 0
                    val range = loanRange(l.startDate, tenure)
                    val cur = l.currency.ifBlank { BASE_CURRENCY }
                    if (l.emiAmount != null) {
                        totalEmiBase += if (cur == BASE_CURRENCY) l.emiAmount else convert(money(l.emiAmount, cur), BASE_CURRENCY, rates(cur, BASE_CURRENCY)).amount
                    }
                    LoanUiModel(
                        id = l.id,
                        lender = l.lender?.takeIf { it.isNotBlank() } ?: "Loan",
                        active = !closed,
                        rangeOrRate = range ?: (l.interestRate?.let { "${it}% p.a." } ?: "—"),
                        paidCountText = if (tenure > 0) "$paid / $tenure paid" else "$paid paid",
                        progress = if (tenure > 0) (paid.toDouble() / tenure.toDouble()).coerceIn(0.0, 1.0) else 0.0,
                        hasTenure = tenure > 0,
                        principalFormatted = formatMoney(l.principal, cur),
                        emiFormatted = l.emiAmount?.let { formatMoney(it, cur) } ?: (if (l.rateType == "variable") "Varies" else "—"),
                    )
                }
                _loans.value = uis
                _totalEmiFormatted.value = formatMoney(totalEmiBase)
            }
        }
    }

    /** Matches web's `AddLoan.save()`: requires a lender name or a
     * principal; EMI is caller-supplied (auto-calculated by the screen
     * for fixed-rate loans via `emiFromPrincipal`, null for variable). */
    suspend fun create(
        lender: String, principalMajorText: String, emiMajorText: String?, interestRateText: String,
        tenureText: String, startDate: String?, dueDayText: String, autoMarkPaid: Boolean, rateType: String,
        fundingAccountId: String?, alertTimeUtc: String,
    ): String? {
        if (lender.trim().isEmpty() && principalMajorText.isBlank()) return "Enter a lender or a loan amount."
        val userId = authRepository.currentUserId.value ?: return "Couldn't determine the current user."
        return try {
            val principalMinor = fromMajor(principalMajorText.toDoubleOrNull() ?: 0.0, BASE_CURRENCY).amount
            val dueDay = dueDayText.toIntOrNull()?.coerceIn(1, 31)
            loansRepository.create(
                userId,
                NewLoanInput(
                    lender = lender.trim(), currency = BASE_CURRENCY, principal = principalMinor,
                    emiAmount = emiMajorText?.toDoubleOrNull()?.let { fromMajor(it, BASE_CURRENCY).amount },
                    interestRate = interestRateText.toDoubleOrNull() ?: 0.0,
                    tenureMonths = tenureText.toIntOrNull(), startDate = startDate, emiDueDay = dueDay,
                    autoMarkPaid = autoMarkPaid, rateType = rateType, fundingAccountId = fundingAccountId,
                    alertTimeUtc = alertTimeUtc,
                ),
            )
            null
        } catch (e: Exception) {
            "Couldn't add the loan: ${e.message}"
        }
    }
}

/** Effective paid-EMI count for a loan row (manual marks union auto-marked
 * past-due) -- matches web's page-local `paidCount()` exactly. */
internal fun paidCount(l: Loan, todayIso: String): Int {
    val tenure = l.tenureMonths ?: 0
    var manual = parseManualPaid(l.emiPayments)
    if (manual.isEmpty() && (l.emisPaid ?: 0) > 0) manual = (1..(l.emisPaid ?: 0)).toList()
    return effectivePaidEmis(manual, tenure, l.autoMarkPaid, l.startDate, l.emiDueDay, todayIso).size
}

/** Parses the `emi_payments` JSON map `{ "1": "2026-01-05", "2": "" }` into
 * the list of EMI numbers with a non-blank paid-on date. */
internal fun parseManualPaid(json: String?): List<Int> {
    if (json.isNullOrBlank()) return emptyList()
    return try {
        val obj = org.json.JSONObject(json)
        obj.keys().asSequence().filter { obj.optString(it).isNotBlank() }.mapNotNull { it.toIntOrNull() }.toList()
    } catch (e: Exception) {
        emptyList()
    }
}

/** "Mar '26 - Nov '26" from a start date + tenure in months -- matches
 * web's `loanRange()`. */
internal fun loanRange(startIso: String?, tenure: Int): String? {
    if (startIso.isNullOrBlank() || tenure <= 0) return null
    return try {
        val start = LocalDate.parse(startIso.take(10))
        val end = start.plusMonths(tenure.toLong())
        val fmt = java.time.format.DateTimeFormatter.ofPattern("MMM ''yy", Locale.ENGLISH)
        "${start.format(fmt)} – ${end.format(fmt)}"
    } catch (e: Exception) {
        null
    }
}

internal fun formatMoney(minor: Long, currency: String = "INR"): String = try {
    NumberFormat.getCurrencyInstance(Locale("en", "IN")).apply {
        this.currency = Currency.getInstance(currency)
        maximumFractionDigits = 0
    }.format(toMajor(money(minor, currency)))
} catch (e: Exception) {
    "$currency ${minor / 100.0}"
}
